include/amici/model.h

```diff
@@ -456,6 +456,35 @@ namespace amici {
         
         void fdwdx(const realtype t, const N_Vector x);
         
+        /** residual function
+         * @param it time index
+         * @param rdata ReturnData instance to which result will be written
+         * @param edata ExpData instance containing observable data
+         */
+        void fres(const int it, ReturnData *rdata, const ExpData *edata);
+        
+        /** chi-squared function
+         * @param it time index
+         * @param rdata ReturnData instance to which result will be written
+         * @param edata ExpData instance containing observable data
+         */
+        void fchi2(const int it, ReturnData *rdata);
+        
+        /** residual sensitivity function
+         * @param it time index
+         * @param rdata ReturnData instance to which result will be written
+         * @param edata ExpData instance containing observable data
+         */
+        void fsres(const int it, ReturnData *rdata, const ExpData *edata);
+        
+        /** fisher information matrix function
+         * @param it time index
+         * @param rdata ReturnData instance to which result will be written
+         * @param edata ExpData instance containing observable data
+         */
+        void fFIM(const int it, ReturnData *rdata);
```

Comment:
Regarding all newly added functions:
None of them requires access to `Model`, or did I miss something? Wouldn't it be better to move them to `ReturnData` or completely free them?

At least all entities not depending on `ExpData` could also be computed on demand to save memory.

Reply:
those that do not directly depend on ExpData indirectly depend on it either through rdata.res or rdata.sres. Yes, this could in theory be moved to ReturnData, but conceptually I think that these computation belong more in the model than ReturnData. In particular if we look at #332, these functions may easily end up with calls to Model->fsy().

Reply:
Hm... okay... `Model` is already so bloated. Maybe that can be cleaned up at a later stage...